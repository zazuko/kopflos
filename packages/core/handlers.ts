import { log, type SubjectHandler } from './index.js'

export function getCoreRepresentation(): SubjectHandler {
  log.warn('Using @kopflos-cms/core/handlers.js#getCoreRepresentation is not equivalent to skipping the handler altogether because it materiliazes the dataset. This is not recommended for large resources.')

  return ({ subject }) => {
    return {
      status: 200,
      body: subject.dataset.toStream(),
    }
  }
}
