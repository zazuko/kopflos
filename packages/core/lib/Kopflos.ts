import type { IncomingHttpHeaders, OutgoingHttpHeaders } from 'node:http'
import type { DatasetCore, NamedNode, Stream } from '@rdfjs/types'
import type { GraphPointer, MultiPointer } from 'clownface'
import type { Options as EndpointOptions, StreamClient } from 'sparql-http-client/StreamClient.js'
import type { ParsingClient } from 'sparql-http-client/ParsingClient.js'
import { CONSTRUCT } from '@tpluscode/sparql-builder'
import { IN } from '@tpluscode/sparql-builder/expressions'
import type { KopflosEnvironment } from './env/index.js'
import { createEnv } from './env/index.js'
import type { ResourceShapeLookup, ResourceShapeMatch } from './resourceShape.js'
import defaultResourceShapeLookup from './resourceShape.js'
import { responseOr } from './responseOr.js'
import type { ResourceLoader, ResourceLoaderLookup } from './resourceLoader.js'
import { fromOwnGraph, findResourceLoader } from './resourceLoader.js'
import type { Handler, HandlerArgs, HandlerLookup } from './handler.js'
import { loadHandler } from './handler.js'
import type { HttpMethod } from './httpMethods.js'

interface KopflosRequest {
  iri: NamedNode
  method: HttpMethod
  headers: IncomingHttpHeaders
}

type ResultBody = Stream | DatasetCore | GraphPointer | Error
export interface ResultEnvelope {
  body?: ResultBody | string
  status?: number
  headers?: OutgoingHttpHeaders
}
export type KopflosResponse = ResultBody | ResultEnvelope

export interface Kopflos {
  get env(): KopflosEnvironment
  get apis(): MultiPointer
  handleRequest(req: KopflosRequest): Promise<ResultEnvelope>
}

interface Clients {
  stream: StreamClient
  parsed: ParsingClient
}

type Endpoint = string | EndpointOptions | Clients

export interface KopflosConfig {
  sparql: Record<string, Endpoint> & { default: Endpoint }
  codeBase?: string
}

export interface Options {
  dataset?: DatasetCore
  resourceShapeLookup?: ResourceShapeLookup
  resourceLoaderLookup?: ResourceLoaderLookup
  handlerLookup?: HandlerLookup
}

type Dataset = ReturnType<KopflosEnvironment['dataset']>

export default class Impl implements Kopflos {
  readonly dataset: Dataset
  readonly env: KopflosEnvironment

  constructor(private readonly config: KopflosConfig, private readonly options: Options = {}) {
    this.env = createEnv(config)

    this.dataset = this.env.dataset([
      ...options.dataset || [],
    ])
  }

  get graph() {
    return this.env.clownface({ dataset: this.dataset })
  }

  get apis(): MultiPointer {
    return this.graph.has(this.env.ns.rdf.type, this.env.ns.kopflos.Api)
  }

  async handleRequest(req: KopflosRequest): Promise<ResultEnvelope> {
    const result = await responseOr(this.findResourceShape(req.iri), (resourceShapeMatch: ResourceShapeMatch) => {
      const resourceShape = this.graph.node(resourceShapeMatch.resourceShape)

      return responseOr(this.findResourceLoader(resourceShape), loader => {
        const coreRepresentation = loader(req.iri, this)

        return responseOr(this.loadHandler(req.method, resourceShapeMatch, coreRepresentation), async handler => {
          const resourceGraph = this.env.clownface({
            dataset: await this.env.dataset().import(coreRepresentation),
          })
          const args: HandlerArgs = {
            resourceShape,
            env: this.env,
            subject: resourceGraph.node(resourceShapeMatch.subject),
            property: undefined,
            object: undefined,
          }
          if ('property' in resourceShapeMatch) {
            args.property = resourceShapeMatch.property
            args.object = resourceGraph.node(resourceShapeMatch.object)
          }

          return handler(args)
        })
      })
    })

    if (this.isEnvelope(result)) {
      return result
    }

    return {
      status: 200,
      body: result,
    }
  }

  async findResourceShape(iri: NamedNode): Promise<ResourceShapeMatch | KopflosResponse> {
    const resourceShapeLookup = this.options.resourceShapeLookup || defaultResourceShapeLookup

    return responseOr(resourceShapeLookup(iri, this), async candidates => {
      if (candidates.length === 0) {
        return { status: 404 }
      }

      if (candidates.length > 1) {
        return new Error('Multiple resource shapes found')
      }

      return candidates[0]
    })
  }

  async findResourceLoader(resourceShape: GraphPointer): Promise<ResourceLoader | KopflosResponse> {
    const resourceLoaderLookup = this.options.resourceLoaderLookup || findResourceLoader

    const loader = await resourceLoaderLookup(resourceShape, this.env)

    return loader || fromOwnGraph
  }

  async loadHandler(method: HttpMethod, resourceShapeMatch: ResourceShapeMatch, coreRepresentation: Stream): Promise<Handler | KopflosResponse> {
    const handlerLookup = this.options.handlerLookup || loadHandler

    const handler = await handlerLookup(resourceShapeMatch, method, this)

    if (handler) {
      return handler
    }

    if (!('property' in resourceShapeMatch) && (method === 'GET' || method === 'HEAD')) {
      return {
        status: 200,
        body: coreRepresentation,
      }
    }

    return { status: 405 }
  }

  private isEnvelope(arg: KopflosResponse): arg is ResultEnvelope {
    return 'body' in arg || 'status' in arg
  }

  static async fromGraphs(kopflos: Impl, ...graphs: Array<NamedNode | string>): Promise<void> {
    const graphsIris = graphs.map(graph => typeof graph === 'string' ? kopflos.env.namedNode(graph) : graph)
    const quads = CONSTRUCT`?s ?p ?o `
      .WHERE`
        GRAPH ?g {
          ?s ?p ?o
        }
        
        FILTER (?g ${IN(...graphsIris)})
      `.execute(kopflos.env.sparql.default.stream)

    for await (const quad of quads) {
      kopflos.dataset.add(quad)
    }
  }
}
