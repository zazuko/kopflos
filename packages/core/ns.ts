import rdf from '@zazuko/env-node'

type Properties = 'api' | 'resourceLoader' | 'handler' | 'method' | 'config'
type Classes = 'Config' | 'Api' | 'ResourceShape' | 'Handler'
type Shorthands = 'DescribeLoader' | 'OwnGraphLoader'

export interface KopflosTerms extends Record<Properties | Classes | Shorthands, never>{
}

export const kl = rdf.namespace<keyof KopflosTerms>('https://kopflos.described.at/')
