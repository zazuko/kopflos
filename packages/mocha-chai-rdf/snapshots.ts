/* eslint-disable @typescript-eslint/no-explicit-any */
import chai, { expect } from 'chai'
import { jestSnapshotPlugin } from 'mocha-chai-jest-snapshot'
import rdf from '@zazuko/env-node'

declare global {
  /* eslint-disable @typescript-eslint/no-namespace */
  namespace Chai {
    interface Assertion {
      canonical: Assertion
    }
  }
}

chai.use(jestSnapshotPlugin())

chai.use((_chai, utils) => {
  const toMatchSnapshot = (chai.Assertion.prototype as any).toMatchSnapshot

  utils.addMethod(chai.Assertion.prototype, 'toMatchSnapshot', function (this: Chai.Assertion, ...args: any[]) {
    const obj = (this as any)._obj

    if (utils.flag(this, 'dataset-canonical')) {
      // Custom behavior for dataset
      expect(rdf.dataset.toCanonical(obj)).toMatchSnapshot(...args)
    } else {
      // Original behavior
      toMatchSnapshot.apply(this, args)
    }
  })

  utils.addProperty(chai.Assertion.prototype, 'canonical', function (this: Chai.Assertion) {
    utils.flag(this, 'dataset-canonical', true)
  })
})
