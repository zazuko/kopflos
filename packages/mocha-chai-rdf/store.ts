/* eslint-disable no-console,camelcase */
import * as url from 'node:url'
import assert from 'node:assert'
import * as Oxigraph from 'oxigraph'
import type { NamespaceBuilder } from '@rdfjs/namespace'
import rdf from '@zazuko/env-node'
import type { DatasetCore, NamedNode, Quad_Graph } from '@rdfjs/types'
import type { AnyPointer } from 'clownface'
import type { Dataset } from '@zazuko/env/lib/Dataset.js'

declare module 'mocha' {
  interface Context {
    dataset: DatasetCore
    graph: AnyPointer
    store: Oxigraph.Store
  }
}

interface DatasetSourceOptions {
  format?: 'trig' | 'nq'
  loadAll?: boolean
  includeDefaultGraph?: boolean
}

interface GraphSourceOptions {
  format?: 'ttl' | 'nt'
}

type Options = (DatasetSourceOptions | GraphSourceOptions) & {
  baseIri?: string | NamespaceBuilder
}

export function createStore(base: string, options: Options = {}) {
  const format = options.format ?? 'ttl'
  let loadAll = true
  let includeDefaultGraph = false
  if (options.format === 'trig' || options.format === 'nq') {
    loadAll = options.loadAll || false
    includeDefaultGraph = options.includeDefaultGraph || false
  }

  return async function (this: Mocha.Context) {
    const store = new Oxigraph.Store()

    const path = url.fileURLToPath(new url.URL(`${base}.${format}`))

    let dataset: Dataset = await rdf.dataset().import(rdf.fromFile(path))

    let graph: Quad_Graph | undefined
    if (this.currentTest && !loadAll) {
      graph = testGraph(this.currentTest)
      dataset = dataset
        .filter((quad) => {
          return quad.graph.equals(graph) || (includeDefaultGraph && quad.graph.equals(rdf.defaultGraph()))
        })
        .map(function toDefaultGraph(quad) {
          return rdf.quad(quad.subject, quad.predicate, quad.object)
        })
    }

    for (const quad of dataset) {
      store.add(quad)
    }

    function assertNotEmpty() {
      if (loadAll) {
        return assert(dataset.size > 0, 'Test data not found')
      }
      if (graph && !graph.equals(rdf.defaultGraph())) {
        return assert(
          dataset.size > 0,
          `Test data not found in GRAPH <${graph.value}>`,
        )
      }

      return assert(
        dataset.match(null, null, null, graph).size > 0,
        `Test data not found in GRAPH <${graph?.value}>`,
      )
    }

    Object.defineProperty(this, 'dataset', {
      get() {
        assertNotEmpty()
        return dataset
      },
      configurable: true,
    })

    Object.defineProperty(this, 'graph', {
      get() {
        assertNotEmpty()
        return rdf.clownface({ dataset })
      },
      configurable: true,
    })
    Object.defineProperty(this, 'store', {
      get() {
        assertNotEmpty()
        return store
      },
      configurable: true,
    })

    // hack to remove the graph after the test just at the right moment
    const cleanup = this.test?.title.includes('before all') ? after : afterEach
    cleanup(() => {
      /* eslint-disable @typescript-eslint/ban-ts-comment */
      // @ts-ignore
      delete this.dataset
      // @ts-ignore
      delete this.graph
      // @ts-ignore
      delete this.store
    })
  }
}

function testGraph(test: Mocha.Test): NamedNode {
  return rdf.namedNode(encodeURI(test.titlePath().slice(1, -1).map(removeSpaces).join('/')))
}

function removeSpaces(arg: string) {
  return arg.replaceAll(' ', '-')
}
