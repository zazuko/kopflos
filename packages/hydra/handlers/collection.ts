import { PassThrough } from 'node:stream'
import type { Handler } from '@kopflos-cms/core'
import { constructQuery } from '@hydrofoil/shape-to-query'
import constraints from '@hydrofoil/shape-to-query/constraints.js'
// eslint-disable-next-line import/no-unresolved
import { kl } from '@kopflos-cms/core/ns.js'
import { memberQueryShape, totalsQueryShape } from '../lib/queryShapes.js'
import { HydraMemberAssertionConstraint } from '../lib/shaclConstraint/HydraMemberAssertionConstraint.js'

constraints.set(kl.HydraMemberAssertionConstraintComponent, HydraMemberAssertionConstraint)

export function get(): Handler {
  return ({ env, subject }) => {
    const body = new PassThrough({ objectMode: true })

    const memberQuery = constructQuery(memberQueryShape({ env, collection: subject }))
    env.sparql.default.stream.query.construct(memberQuery).pipe(body)

    const totalQuery = constructQuery(totalsQueryShape({ env, collection: subject }))
    env.sparql.default.stream.query.construct(totalQuery).pipe(body)

    return {
      status: 200,
      body,
    }
  }
}
