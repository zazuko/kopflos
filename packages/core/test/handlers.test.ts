import type { Stream } from '@rdfjs/types'
import { expect, use } from 'chai'
import rdf from '@zazuko/env-node'
import snapshots from 'mocha-chai-rdf/snapshots.js'
import { createStore } from 'mocha-chai-rdf/store.js'
import { getCoreRepresentation } from '../handlers.js'
import type { Body, KopflosConfig } from '../lib/Kopflos.js'
import { ex } from '../../testing-helpers/ns.js'
import Kopflos from '../lib/Kopflos.js'
import handlerLookupStub from './support/handlerLookupStub.js'

describe('@kopflos-cms/handlers.js', function () {
  use(snapshots)

  before(createStore(import.meta.url, { format: 'trig', includeDefaultGraph: true }))

  const config: KopflosConfig = {
    baseIri: 'http://example.com/',
    sparql: {
      default: 'http://localhost:8080/sparql',
    },
  }

  describe('getCoreRepresentation', function () {
    it('forwards core representation', async function () {
      // given
      const kopflos = new Kopflos(config, {
        dataset: this.rdf.dataset,
        resourceShapeLookup: async () => [{
          api: ex.api,
          resourceShape: ex.FooShape,
          subject: ex.foo,
        }],
        handlerLookup: handlerLookupStub(getCoreRepresentation()),
      })

      // when
      const response = await kopflos.handleRequest({
        iri: ex.foo,
        method: 'GET',
        headers: {},
        body: {} as Body,
        query: {},
      }) as unknown as { status: number; body: Stream }

      // then
      expect(response).to.have.property('status', 200)
      const dataset = await rdf.dataset().import(response.body)
      expect(dataset).canonical.toMatchSnapshot()
    })
  })
})
