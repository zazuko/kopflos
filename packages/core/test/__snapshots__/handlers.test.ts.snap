// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@kopflos-cms/handlers.js getCoreRepresentation forwards core representation 1`] = `
"<http://example.org/foo> <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://schema.org/Thing> .
"
`;
